x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java

```diff
@@ -0,0 +1,121 @@
+ */
+class TopMetricsAggregator extends MetricsAggregator {
+    private final BucketedSort sort;
+    private final String metricName;
+    private final ValuesSource.Numeric metricValueSource;
+    private DoubleArray values;
+
+    TopMetricsAggregator(String name, SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators,
+            Map<String, Object> metaData, BucketedSort sort,
+            String metricName, ValuesSource.Numeric metricValueSource) throws IOException {
+        super(name, context, parent, pipelineAggregators, metaData);
+        this.sort = sort;
+        this.metricName = metricName;
+        this.metricValueSource = metricValueSource;
+        if (metricValueSource != null) {
+            values = context.bigArrays().newDoubleArray(2, false);
+            values.fill(0, values.size(), Double.NaN);
+        }
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        boolean needs = (sort != null && sort.needsScores()) || (metricValueSource != null && metricValueSource.needsScores());
+        return needs ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        assert sub == LeafBucketCollector.NO_OP_COLLECTOR : "Expected noop but was " + sub.toString();
+
+        if (metricValueSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        BucketedSort.Leaf leafSort = sort.forLeaf(ctx);
+        // TODO allow configuration of value mode
+        NumericDoubleValues metricValues = MultiValueMode.AVG.select(metricValueSource.doubleValues(ctx));
+
+        return new LeafBucketCollector() { // TODO do we need to extend *Base*? It doesn't look like we use it.
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                if (leafSort.collectIfCompetitive(doc, bucket)) {
+                    if (bucket >= values.size()) {
+                        long oldSize = values.size();
+                        values = context.bigArrays().grow(values, bucket + 1);
+                        values.fill(oldSize, values.size(), Double.NaN);
+                    }
+                    double metricValue = metricValues.advanceExact(doc) ? metricValues.doubleValue() : Double.NaN; 
+                    values.set(bucket, metricValue);
+                }
+            }
+
+            @Override
+            public void setScorer(Scorable s) throws IOException {
+                leafSort.setScorer(s);
+            }
+        };
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long bucket) throws IOException {
+        if (metricValueSource == null) {
+            return buildEmptyAggregation();
+        }
+        double metricValue = values.get(bucket);
+        SortValue sortValue = sort.getValue(bucket);
```

Comment:
This is much cleaner, thanks for rewiring it to send the SortValue (despite making that class comparatively a lot larger b/c of all the serialization)